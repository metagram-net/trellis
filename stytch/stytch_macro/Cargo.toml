[package]
name = "stytch_macro"
version = "0.1.0"
edition = "2018"

[lib]
proc-macro = true

[dependencies]
syn = { version = "1.0", features = ["extra-traits", "full", "printing"]}
quote = "1.0"

[dev-dependencies]
http = "0.2"
serde_json = "1.0"
serde = { version = "1.0", features = ["derive"] }
stytch = { path = ".." }
trybuild = "1.0"
