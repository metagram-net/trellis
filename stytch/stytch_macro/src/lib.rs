extern crate proc_macro;

use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, AttributeArgs, ItemStruct, Lit, NestedMeta};

#[proc_macro_attribute]
pub fn post(attr: TokenStream, item: TokenStream) -> TokenStream {
    // TODO: I wish I didn't have to clone this just for error messages :\
    let args = attr.clone();
    let attr_ast = parse_macro_input!(attr as AttributeArgs);

    let route = match &attr_ast[..] {
        [NestedMeta::Lit(Lit::Str(r))] => r,
        _ => panic!("expected 1 string argument, got [{}]", args),
    };

    let item_ast = parse_macro_input!(item as ItemStruct);

    // Build the trait implementation
    impl_post(route, &item_ast)
}

fn impl_post(route: &syn::LitStr, item: &ItemStruct) -> TokenStream {
    let struct_name = &item.ident;
    let generics = &item.generics;
    let gen = quote! {
        #item
        impl #generics ::stytch::endpoint::Endpoint for #struct_name #generics {
            fn method(&self) -> ::http::Method {
                ::http::Method::POST
            }
            fn path(&self) -> ::std::string::String {
                ::std::string::String::from(#route)
            }
            fn query(&self) -> ::stytch::endpoint::Query {
                ::stytch::endpoint::Query::default()
            }
            fn body(&self) -> ::serde_json::Result<::stytch::endpoint::Body> {
                ::stytch::endpoint::Body::from_data(self)
            }
        }
    };
    gen.into()
}
