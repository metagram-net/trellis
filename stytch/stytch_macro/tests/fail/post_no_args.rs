use serde::Serialize;
use stytch_macro::post;

#[post()]
#[derive(Serialize)]
struct PostNoArgs<'a> {
    id: &'a str,
}

fn main() {}
