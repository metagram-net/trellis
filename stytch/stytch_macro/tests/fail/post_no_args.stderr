error: custom attribute panicked
 --> $DIR/post_no_args.rs:4:1
  |
4 | #[post()]
  | ^^^^^^^^^
  |
  = help: message: expected 1 string argument, got []
